src/Pact/Native/Decrypt.hs

```diff
@@ -0,0 +1,216 @@
+{-# LANGUAGE OverloadedStrings #-}
+-- |
+-- Module      :  Pact.Native.Decrypt
+-- Copyright   :  (C) 2019 Stuart Popejoy
+-- License     :  BSD-style (see the file LICENSE)
+-- Maintainer  :  Stuart Popejoy <stuart@kadena.io>
+--
+-- Built-ins for decrypting commitments.
+--
+
+
+module Pact.Native.Decrypt
+  ( decryptDefs
+#if !defined(ghcjs_HOST_OS)
+  , doEncrypt
+#endif
+  ) where
+
+import Control.Monad
+import qualified Data.ByteArray as BA
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import Data.Default (def)
+import Data.Text (Text)
+import Data.Text.Encoding
+
+import Pact.Native.Internal
+import Pact.Types.Pretty
+import Pact.Types.Runtime
+
+#if !defined(ghcjs_HOST_OS)
+import Crypto.PubKey.Curve25519
+import Crypto.Error
+import Crypto.Cipher.ChaChaPoly1305
+import Crypto.MAC.Poly1305 (Auth(..),authTag)
+#else
+
+#endif
+
+
+decryptDefs :: NativeModule
+decryptDefs = ("Decryption",
+  [ validateKeypairDef
+  , decryptDef
+  ])
+
+validateKeypairDef :: NativeDef
+validateKeypairDef =
+  defRNative "validate-keypair"
+  validateKeypair
+  (funType tTyBool [("public",tTyString),("secret",tTyString)])
+  [LitExample "(validate-keypair pubkey privkey)"]
+  ("Enforce that the Curve25519 keypair of (PUBLIC,SECRET) match. " <>
+   "Key values are base-16 strings of length 32.")
+  where
+    validateKeypair :: RNativeFun e
+    validateKeypair i [p@(TLitString p'), s@(TLitString s')] = do
+      public <- doPublic p p'
+      secret <- doSecret s s'
+      v <- doValidate i public secret
+      if v then return $ toTerm v
+        else evalError' i $ "validate-keypair: mismatch"
+    validateKeypair i as = argsError i as
+
+doValidate :: HasInfo i => i -> ByteString -> ByteString -> Eval e Bool
+#if defined(ghcjs_HOST_OS)
+doValidate i _ _ = evalError' i "'validate-keyset' unsupported in javascript-backed Pact"
+#else
+
+doValidate i public secret = do
+  pk <- importPublic i public
+  sk <- importSecret i secret
+  return $ toPublic sk == pk
+
+#endif
+
+
+
+decryptDef :: NativeDef
+decryptDef =
+  defRNative "decrypt-cc20p1305"
+  decrypt'
+  (funType tTyString [("ciphertext",tTyString),("nonce",tTyString),("aad",tTyString),
+                      ("public-key",tTyString),("secret-key",tTyString)])
+  [LitExample "(decrypt-cc20p1305 ciphertext nonce aad pubkey privkey)"]
+  ("Decrypt a Curve25519/ChaCha20/Poly1305 CIPHERTEXT using NONCE, AAD, PUBLIC-KEY and SECRET-KEY. " <>
+   "CIPHERTEXT is an unpadded base64url value with the 16-byte authentication tag output of the encryption. " <>
+   "PUBLIC-KEY and SECRET-KEY are base-16 strings of length 32. " <>
+   "NONCE is an unpadded base64URL value of a 24-byte bytestring. " <>
+   "AAD is additional authentication data of an unpadded base-16 string of any length. " <>
+   "Result is unpadded base64URL.")
+  where
+    decrypt' :: RNativeFun e
+    decrypt' i [c@(TLitString c'),n@(TLitString n'),a@(TLitString a'),p@(TLitString p'),s@(TLitString s')] = do
```

Comment:
I branched this to see if the info given by the `FunApp` would differ significantly from passing `c/n/a/p/s` as the info object of choice. It gives the same positional info, so these patterns may be redundant - we can just use `i` when calling `do-*` here and above

Reply:
Are you sure about that? The whole reason is to point at the argument itself. Can you paste your output showing that the positional info doesn't differ?